### PR TITLE
Added journal compaction behavior for LevelDB #21677

### PR DESCRIPTION
**Pull request by nick-nachos**
Added journal compaction behavior for the **local** LevelDB plugin. The underlying issue was discussed [here](https://github.com/akka/akka/issues/21677#issuecomment-312786915)

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
Hi @nick-nachos,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by raboof** [comment]
OK TO TEST

File: akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala
Comment (raboof) [inline akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: switching with an `isInstanceOf` here doesn't feel great, but I'm not sure of an alternative... is compaction not applicable to `SharedLeveldbStore`? (then we could mix in `LeveldbCompaction` directly here and skip the check, right?)

File: akka-persistence/src/test/scala/akka/persistence/journal/leveldb/JournalCompactionSpec.scala
Comment (raboof) [inline akka-persistence/src/test/scala/akka/persistence/journal/leveldb/JournalCompactionSpec.scala]: Perhaps throw?

File: akka-docs/src/test/scala/docs/persistence/PersistencePluginDocSpec.scala
Comment (raboof) [inline akka-docs/src/test/scala/docs/persistence/PersistencePluginDocSpec.scala]: Configuring this per-`persistenceId` is pretty low-level... though I guess we could allow more broad-brush configuration by supporting wildcards here later.

**Review by raboof — COMMENTED** [review]
Thanks for putting this together! I'm not sure yet about the structuring, nicely testable though. I've added some review comments.

File: akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbCompaction.scala
Comment (nick-nachos) [inline akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbCompaction.scala]: I left these comments there on purpose actually for us to discuss this. I actually wanted to add some logging here, but I noticed that ActorLogging was not already mixed into LeveldbJournal, so I was reluctant to do so. Do you confirm that it is OK for me to mix it in?

File: akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbCompaction.scala
Comment (nick-nachos) [inline akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbCompaction.scala]: Indeed it was, I missed it somehow, will fix.

File: akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala
Comment (nick-nachos) [inline akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: I had initially mixed the LeveldbCompaction trait into LeveldbStore, however when I run the validatePullRequest task I got a "binary compatibility" error, since SharedLeveldbStore (which mixes in LeveldbStore) is actually a public class. Given the fact that this does not seem to be a really important feature to incorporate into the shared journal (which is only used for testing purposes), I chose to mix LeveldbCompaction into LeveldbJournal (which is package protected) thus conforming with the binary compatibility check. However, given that the compaction functionality is dependent to the sequence number of the last message being deleted, it is necessary for LeveldbStore to trigger the process and pass this info to it, hence the cast was required to do the job. Of course, I am open to suggestions :D

File: akka-persistence/src/test/scala/akka/persistence/journal/leveldb/JournalCompactionSpec.scala
Comment (nick-nachos) [inline akka-persistence/src/test/scala/akka/persistence/journal/leveldb/JournalCompactionSpec.scala]: This was actually a workaround because "calculateJournalSize" is used in the tests to calculate the size of the directory prior to its creation as well (the journal directory is created once the persistent actor is deployed). I could just initialize the journal size variable with zero, but I just wanted to use "calculateJournalSize" in a more uniform manner. This implementation would not cause any issue to the tests (e.g. false positives) since if the journal directory is not created due to any transient failure, most tests will fail anyway. However, I can obviously change this implementation and initialize the variables directly with zero, no prob.

File: akka-docs/src/test/scala/docs/persistence/PersistencePluginDocSpec.scala
Comment (nick-nachos) [inline akka-docs/src/test/scala/docs/persistence/PersistencePluginDocSpec.scala]: I thought of the possibility of wildcards as well, but that is a more complex issue than what it initially seems. If you check the signature and documentation of the **compactRange** method of the LevelDB API, you will see that the start-end parameters are byte arrays, which are calculated by the LeveldbStore using the Key utility object (akka api). The Key object in turn uses the persistence ids to calculate the most significant bytes of these arrays, thus implying that the messages of each distinct persistence-id are stored in a continuous range, which is mutually exclusive to other ranges (obviously for performance reasons). This however is the very reason that a wildcard-based compaction could problematic: it would probably lead to performing compaction across multiple ranges, which would effectively try to compact undeleted messages as well. This makes no sense as compaction does nothing when it comes to messages without tombstones, but I have the feeling that due to the much wider range of the scan, it would come at a significant performance cost. I think this may be what is implied by the documentation string of the compactRange method found [here](https://github.com/google/leveldb/blob/master/include/leveldb/db.h) which says: 

> This operation should typically only be invoked by users who understand the underlying implementation.

Given the thoughts above, I kind of dropped the wildcard though for now.

**Comment by lightbend-cla-validator** [comment]
Hi @invalid-email-address,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

File: akka-docs/src/test/scala/docs/persistence/PersistencePluginDocSpec.scala
Comment (patriknw) [inline akka-docs/src/test/scala/docs/persistence/PersistencePluginDocSpec.scala]: might be better with real values here instead of comments
```
persistence-id-1 = 30s
persistence-id-2 = 60s
```

File: akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala
Comment (patriknw) [reply akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: `LeveldbStore` is also internal api, so you are free to change it, and add mima exclusions to src/main/mima-filters/

File: akka-docs/src/test/scala/docs/persistence/PersistencePluginDocSpec.scala
Comment (patriknw) [reply akka-docs/src/test/scala/docs/persistence/PersistencePluginDocSpec.scala]: I agree with @raboof. Would be better with one single property for all persistenceIds. I don't understand your reasoning. The compaction is triggered by a delete, so the persistenceId to is know.

Comment (patriknw) [reply akka-docs/src/test/scala/docs/persistence/PersistencePluginDocSpec.scala]: all persistenceIds are typically not know up front, i.e. it would be difficult to configure them like this

**Review by patriknw — CHANGES_REQUESTED** [review]
The feature is looking good. Would like to simplify the configuration.

File: akka-docs/src/test/scala/docs/persistence/PersistencePluginDocSpec.scala
Comment (nick-nachos) [reply akka-docs/src/test/scala/docs/persistence/PersistencePluginDocSpec.scala]: Actually the intervals are message intervals not time ones, which actually proves the point of putting some actual values to the documentation. Will fix

File: akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala
Comment (nick-nachos) [reply akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: [SharedLeveldbStore](https://github.com/akka/akka/blob/master/akka-persistence/src/main/scala/akka/persistence/journal/leveldb/SharedLeveldbStore.scala) which extends LeveldbStore though is public, that is why I got the original binary compatibility flag. I assume these mima exclusions allow me to work around this, right?

Comment (patriknw) [reply akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: ok, SharedLeveldbStore is not intended for subclassing. It should have been declared final. It's fine to make that exclusion.

File: akka-docs/src/test/scala/docs/persistence/PersistencePluginDocSpec.scala
Comment (nick-nachos) [reply akka-docs/src/test/scala/docs/persistence/PersistencePluginDocSpec.scala]: @patriknw when I initially thought of the wildcard feature, I got the perspective of doing a single compaction across all persistence ids, which would have led to the issues described above. I now get that the suggestion was to just use the wildcard to dynamically apply the same interval to perform separate compactions for each persistence id that has not been specifically defined in the configuration. I will proceed with fixing this right away

Comment (patriknw) [reply akka-docs/src/test/scala/docs/persistence/PersistencePluginDocSpec.scala]: good, I don't think we need to be able to specify different values for different persistenceIds at all, one single property `compaction-interval` would be enough, unless you have a specific need for more fine grained configuration yourself

Comment (nick-nachos) [reply akka-docs/src/test/scala/docs/persistence/PersistencePluginDocSpec.scala]: The use case I dealt with at work actually requires a more fine grained configuration, so I would prefer to keep that, if that is not an issue with you.

File: akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala
Comment (nick-nachos) [reply akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: @patriknw one question before making the commit, given I'm not familiar with mima filters: I have put the filters suggested by the mimaReportBinaryIssues task in a file named _2.5.4.backwards.excludes_ and that seems to do the trick. Would that be an acceptable way to go?

File: akka-docs/src/test/scala/docs/persistence/PersistencePluginDocSpec.scala
Comment (patriknw) [reply akka-docs/src/test/scala/docs/persistence/PersistencePluginDocSpec.scala]: that is fine, thanks for contributing

File: akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala
Comment (patriknw) [reply akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: yes, that's the way :)

**Comment by nick-nachos** [comment]
@patriknw, there's a *Close and comment* button on my UI for some reason, is this expected? (there used to be a message indicating that *only members with write access to the repo can merge*). Either way, I will await for a final review to confirm that the final changes produced an acceptable delivery :)

**Comment by patriknw** [comment]
Thanks, we'll review the changes

File: akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala
Comment (raboof) [inline akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: perhaps could be something like `obj.unwrapped().mapValues(v => java.long.parseLong(v.toString)))`?

Comment (nick-nachos) [reply akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: According to the scala docs 'mapValues' returns a view instead of a concrete map (confirmed it with a code sample). Hence, should we go with something like this instead?
`obj.unwrapped().transform((_, v) => java.long.parseLong(v.toString))`

Comment (nick-nachos) [reply akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: @raboof I would prefer a concrete collection instead of view, as it is used after every single delete, multiple times

File: akka-persistence/src/main/mima-filters/2.5.4.backwards.excludes
Comment (patriknw) [inline akka-persistence/src/main/mima-filters/2.5.4.backwards.excludes]: Should be fine to remove the ActorLogging entries, since nothing was changed there

File: akka-persistence/src/main/mima-filters/2.5.4.backwards.excludes
Comment (patriknw) [inline akka-persistence/src/main/mima-filters/2.5.4.backwards.excludes]: isn't `CompactionSegmentManagement` a new class? then it should not be needed here

File: akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala
Comment (patriknw) [reply akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: yes, we shouldn't use views

Comment (patriknw) [reply akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: please use the more explicit version:
```
import scala.collection.JavaConverters._
obj.unwrapped().asScala.map
```

**Review by patriknw — CHANGES_REQUESTED** [review]
one thing about implicit conversions, and then look over the mima filters

File: akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala
Comment (nick-nachos) [reply akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala]: Agreed. I will keep the mapping process as is (```map(entry ⇒ (entry._1, java.lang.Long.parseLong(entry._2.toString)))```), since this allows for using both long and string literals in the configuration (i.e. both 4 and "4" will work just fine)

File: akka-persistence/src/main/mima-filters/2.5.4.backwards.excludes
Comment (nick-nachos) [reply akka-persistence/src/main/mima-filters/2.5.4.backwards.excludes]: Unfortunately @patriknw this (along with the *ActorLogging* issue bellow) is a more complex issue than it seems, and I will absolutely need your feedback on how to deal with it. These filters are mandated by the ```mimaReportBinaryIssues``` SBT task, otherwise binary compatibility errors are raised. The *InheritedNewAbstractMethodProblem* is raised due to the fact that *LeveldbStore* now mixes in the *CompactionSegmentManagement* and *ActorLogging*, thus inheriting the corresponding members from them. The thing is however that there seems to be a problematic approach in the implementation of the mima tool regarding this issue (had to download the mima sources to figure it out): If you check the [problems implementation](https://github.com/typesafehub/migration-manager/blob/master/core/src/main/scala/com/typesafe/tools/mima/core/Problems.scala) you will notice that the *InheritedNewAbstractMethodProblem* does include the class info of the class that inherits an abstract member, however that class info is only used to compose the **description** of the problem, and is not used to somehow override the **matchName**, which is the property used for matching by the [ProblemFilters](https://github.com/typesafehub/migration-manager/blob/master/core/src/main/scala/com/typesafe/tools/mima/core/Filters.scala) predicate. The implementation of **matchName** inherited via *AbstractMethodProblem* -> *MethodProblem* only uses the info of the method which introduces the problem, whose fully qualified name in the case of an inherited method is based on the original class/trait that introduces it (hence the misinterpretation one may end up to, thinking that the filter is about the class/trait that introduces the method, instead of the one inheriting it). 
The problem with the above is that, event though these filters actually do the job in this case, they are in fact **way too broad**. The fact that the filters fail to take the inheriting class info into account can lead to ignoring _ActorLogging_ mix-ins from any class/trait into the akka-persistence module, and that would result into false-negatives into binary compatibility checks. On the other hand not adding these filters produces binary compatibility errors which will lead to failure of the ```validatePullRequest``` SBT task. Thus the questing is, how do we proceed?

Comment (patriknw) [reply akka-persistence/src/main/mima-filters/2.5.4.backwards.excludes]: I see. That is strange and slightly dangerous because after this it would not detect new breakages if ActorLogging was added to something else. I have created issue https://github.com/typesafehub/migration-manager/issues/196

Until that has been fixed we have to include these filters.

**Comment by nick-nachos** [comment]
Fixed the issue with the implicit conversion, as requested.

**Review by patriknw — APPROVED** [review]
LGTM

**Comment by patriknw** [comment]
@raboof could you take another look since you reviewed previously?

**Review by raboof — APPROVED** [review]
Seems safe, thanks!

**Comment by patriknw** [comment]
Refs #21677

**Pull request merged**
